### PR TITLE
Add fault-tolerant custom source

### PR DESCRIPTION
**Pull request by viliam-durina**

### CODE REVIEW

**Comment by Holmistr** [comment]
verify

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java [hunk 1]
```diff
@@ -26,46 +29,54 @@
...
 
         /**
-         * Closes the buffer, signaling that all items have been emitted.
+         * Closes the buffer, signaling that all items have been emitted. Only
+         * {@linkplain #batch} sources are allowed to call this method.
+         *
+         * @throws JetException if the source is a streaming source
          */
-        void close();
+        void close() throws JetException;
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java]: this is a breaking change

Comment (mtopolnik) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/SourceBuilder.java]: `JetException` is unchecked, is it really a breaking change?

**Comment by viliam-durina** [comment]
Summary of breaking changes:
- in Pipeline API, streaming custom source now doesn't allow `buffer.close()` method
- in Core API, the `SourceProcessors` the `convenientSourceP` and `convenientTimestampedSourceP` have changed signature.

**Pull request merged**